### PR TITLE
Use a 'Greedy' approach for aggregations on measurement.

### PR DESCRIPTION
**Pull request by songy23**

### CODE REVIEW

File: core_impl/src/test/java/com/google/instrumentation/stats/StatsManagerImplTest.java [hunk 1]
```diff
@@ -192,16 +199,26 @@ public void testRecordNonExistentMeasurementDescriptor() {
...
         (DistributionView) statsManager.getView(RpcConstants.RPC_CLIENT_ROUNDTRIP_LATENCY_VIEW);
-    // Won't record stats if there are non existent tags.
-    assertThat(view.getDistributionAggregations()).hasSize(0);
+
+    DistributionAggregation distributionAggregation = view.getDistributionAggregations().get(0);
+    assertThat(view.getDistributionAggregations()).hasSize(1);
+    List<Tag> tags = distributionAggregation.getTags();
+    // Won't record the unregistered tag key, will use default tag instead:
+    // "method" : "unknown/not set".
+    assertThat(tags.get(0).getKey()).isNotEqualTo(wrongTagKey);
```
Comment (sebright) [inline core_impl/src/test/java/com/google/instrumentation/stats/StatsManagerImplTest.java]: > assertThat(tags.get(0).getKey()).isNotEqualTo(wrongTagKey);

I think this assertion is redundant when we also have `assertThat(tags.get(0).getKey()).isEqualTo(tagKey);`.

File: core_impl/src/test/java/com/google/instrumentation/stats/StatsManagerImplTest.java [hunk 2]
```diff
@@ -84,16 +85,24 @@ public void testGetNonexistentView() throws Exception {
   @Test
   public void testRecord() {
     statsManager.registerView(RpcConstants.RPC_CLIENT_ROUNDTRIP_LATENCY_VIEW);
+    Timestamp start = Timestamp.fromMillis(System.currentTimeMillis());
     for (double val : Arrays.<Double>asList(10.0, 20.0, 30.0, 40.0)) {
       statsManager.record(
           oneTag, MeasurementMap.of(RpcConstants.RPC_CLIENT_ROUNDTRIP_LATENCY, val));
+      // Simulate transmission delay.
+      try {
+        Thread.sleep(500);
```
Comment (sebright) [inline core_impl/src/test/java/com/google/instrumentation/stats/StatsManagerImplTest.java]: We should use a fake clock instead, to keep the tests deterministic and fast.  Also, I think that any refactoring to use a fake clock should be done in a separate PR, because it would be a large change.

File: core_impl/src/test/java/com/google/instrumentation/stats/StatsManagerImplTest.java [hunk 3]
```diff
@@ -192,16 +199,26 @@ public void testRecordNonExistentMeasurementDescriptor() {
...
+
   @Test
   public void testRecordNonExistentTag() {
     statsManager.registerView(RpcConstants.RPC_CLIENT_ROUNDTRIP_LATENCY_VIEW);
-    statsManager.record(
-        wrongTag, MeasurementMap.of(RpcConstants.RPC_CLIENT_ROUNDTRIP_LATENCY, 10.0));
+    statsManager.record(wrongTag, MeasurementMap.of(
+        RpcConstants.RPC_CLIENT_ROUNDTRIP_LATENCY, 10.0));
+    statsManager.record(wrongTag, MeasurementMap.of(
+        RpcConstants.RPC_CLIENT_ROUNDTRIP_LATENCY, 50.0));
```
Comment (sebright) [inline core_impl/src/test/java/com/google/instrumentation/stats/StatsManagerImplTest.java]: I think it would be useful to check that stats for multiple StatsContexts with non-matching keys are combined.  You could use a second set of tags here or create a new test.

File: core_impl/src/test/java/com/google/instrumentation/stats/StatsManagerImplTest.java [hunk 4]
```diff
@@ -192,16 +199,26 @@ public void testRecordNonExistentMeasurementDescriptor() {
     assertThat(view.getDistributionAggregations()).hasSize(0);
   }
 
-  // TODO(songya): update this test once we determine how to handle tags that aren't an exact match
-  // for the view.
+
   @Test
   public void testRecordNonExistentTag() {
```
Comment (sebright) [inline core_impl/src/test/java/com/google/instrumentation/stats/StatsManagerImplTest.java]: This test's name seems less accurate now.  How about something like testRecord_TagNotMatchingView?

File: core_impl/src/main/java/com/google/instrumentation/stats/MutableView.java [hunk 5]
```diff
@@ -29,6 +30,7 @@
  */
 // TODO(songya): remove or modify the methods of this class, since it's not part of the API.
 abstract class MutableView {
+  @VisibleForTesting static final TagValue UNKNOWN_TAG_VALUE = TagValue.create("unknown/not set");
```
Comment (songy23) [inline core_impl/src/main/java/com/google/instrumentation/stats/MutableView.java]: Marked. The current value is from previous discussion (not sure if we've come to a conclusion on this).

File: core_impl/src/test/java/com/google/instrumentation/stats/StatsManagerImplTest.java (hunk 2 above)
Comment (songy23) [inline core_impl/src/test/java/com/google/instrumentation/stats/StatsManagerImplTest.java]: Makes sense, I'll leave a TODO here.

File: core_impl/src/test/java/com/google/instrumentation/stats/StatsManagerImplTest.java [hunk 6]
```diff
@@ -180,7 +179,15 @@ public void testRecordWithEmptyStatsContext() {
...
     DistributionView view =
         (DistributionView) statsManager.getView(RpcConstants.RPC_CLIENT_ROUNDTRIP_LATENCY_VIEW);
-    assertThat(view.getDistributionAggregations()).hasSize(0);
+    assertThat(view.getDistributionAggregations()).hasSize(1);
+    DistributionAggregation distributionAggregation = view.getDistributionAggregations().get(0);
+    List<Tag> tags = distributionAggregation.getTags();
+    assertThat(tags.get(0).getKey()).isEqualTo(tagKey);
+    // Tag is missing for associated measurementValues, should use default tag value
+    // "unknown/not set".
+    assertThat(tags.get(0).getValue()).isEqualTo(MutableView.UNKNOWN_TAG_VALUE);
```
Comment (songy23) [inline core_impl/src/test/java/com/google/instrumentation/stats/StatsManagerImplTest.java]: Add this test in MutableViewTest.

File: core_impl/src/test/java/com/google/instrumentation/stats/StatsManagerImplTest.java (hunk 4 above)
Comment (songy23) [inline core_impl/src/test/java/com/google/instrumentation/stats/StatsManagerImplTest.java]: Done

File: core_impl/src/test/java/com/google/instrumentation/stats/StatsManagerImplTest.java (hunk 1 above)
Comment (songy23) [inline core_impl/src/test/java/com/google/instrumentation/stats/StatsManagerImplTest.java]: Removed the first one.

File: core_impl/src/test/java/com/google/instrumentation/stats/StatsManagerImplTest.java (hunk 3 above)
Comment (songy23) [inline core_impl/src/test/java/com/google/instrumentation/stats/StatsManagerImplTest.java]: Sure, updated this test.

**Comment by songy23** [comment]
Thanks!
